### ISSUE TITLE
Typescript types for series don't match actual functionality

### ISSUE THREAD

**Issue by brentguistwite** [highcharts/highcharts#12242]
#### Expected behaviour
The Highchart api [docs](https://api.highcharts.com/highcharts/series) suggest that regardless of the series type, the `type` option inside `series` is optional.

This is a quote from the API docs
> "If the type option is not specified, it is inherited from chart.type. In TypeScript the type option must always be set."

This essentially forces the user to either forgo type safety by using `any` or type assertion using `as`, or loop over a potentially huge set of data on the frontend just to add the type to each entry in the series. Neither of which are great solutions.

#### Actual behaviour
The `type` key should be appropriately typed as being an optional key inside `series`. 

#### Live demo with steps to reproduce
N/A

#### Product version

Highcharts version 7.2.0

#### Affected browser(s)
N/A

**Comment by sebastianbochan** [comment]
Hi @brentguistwite,
Thank you for reporting. 

FYI @bre1470

**Comment by bre1470** [comment]
This is by design of TypeScript and will not be solved. Use `Highcharts.SeriesOptionsType` to reference every option instead of `any`.

**Comment by brentguistwite** [comment]
@bre1470 Can you expand on what you mean by "This is by design of Typescript"? I'm not sure I understand the point you are trying to make.

**Comment by bre1470** [comment]
@brentguistwite You have to set a static type string so that TypeScript can identify, which of the series options interfaces it should use to check the option types. The chart type option is outside of the series option scope and therefor dynamic for the series options interface. TypeScript is then not able to identify the series options during transpiling, as the chart type option could change on runtime and make any validation for common series options during transpiling invalid.

**Comment by brentguistwite** [comment]
> " Use Highcharts.SeriesOptionsType to reference every series options instead of any."

This doesn't seem to solve the issue, as its expecting the key `type` to be present on each item in the series. This is the issue that I was bringing up with this card. If you don't have the `type` key on every item in your series then TypeScript wont compile. As I mentioned before, the solution to this seems to be that you can either
a) Loop your entire data set, adding the type to every item.
b) Type series as `any`.
c) Use type assertion with `as`.
None of these seem like great options.

This seems like more of an issue with how the code is implemented as opposed to this being a TypeScript issue. My point with opening this issue is that the types are wrong. The type key is NOT a required field because the code works without it. I looked through the types for a bit and the chart type as far as I can tell is just a string. I don't see why this can't be typed as not being required. Sure, you probably would need to do some checks somewhere to make sure that when the time comes it not actually undefined. With the amount of `as any`'s in the codebase it seems like this could easily be fixed by following that pattern.

**Comment by bre1470** [comment]
@brentguistwite Use `Partial<Highcharts.SeriesOptionsType>` instead of `Highcharts.SeriesOptionsType`, so that every property becomes optional.

**Comment by alumni** [comment]
I also agree that the way Highcharts defines types for Series Options is not the best. There are a few things wrong:

1. What you just mentioned above, when you add a series to the chart, not all options should be required, as per the docs (and actual behavior).
2. We should be able to add custom fields (e.g. that we can later in Tooltips).
3. You shouldn't use a Type Union, it makes it impossible to extend Highcharts and create our own custom series. This TS issue explains best how to define the type in such cases: https://github.com/microsoft/TypeScript/issues/28078#issuecomment-432339564

**Comment by bre1470** [comment]
Hello, @alumni!
Thank you for your input.

> I also agree that the way Highcharts defines types for Series Options is not the best.

We are aware of the implied problem with the current approach and constantly revaluate alternatives like option classes and feature sets. 

> 1. What you just mentioned above, when you add a series to the chart, not all options should be required, as per the docs (and actual behavior).

Yes, though with `Partial<T>`only  the `type` option becomes optional. The rest is already optional.

> 2. We should be able to add custom fields (e.g. that we can later in Tooltips).

We describe in our general [documentation](https://www.highcharts.com/docs/advanced-chart-features/highcharts-typescript-declarations#extending-highcharts-in-typescript), how to add custom properties to existing Highcharts types.

> 3. You shouldn't use a Type Union, it makes it impossible to extend Highcharts and create our own custom series. This TS issue explains best how to define the type in such cases: [microsoft/TypeScript#28078 (comment)](https://github.com/microsoft/TypeScript/issues/28078#issuecomment-432339564)

Agreed. We will try the outlined approach ASAP to make the introduction of custom series easier. Reference: #13374

**Comment by alumni** [comment]
> Yes, though with `Partial<T>`only the `type` option becomes optional. The rest is already optional.

> We describe in our general [documentation](https://www.highcharts.com/docs/advanced-chart-features/highcharts-typescript-declarations#extending-highcharts-in-typescript), how to add custom properties to existing Highcharts types.

Actually, our workaround is something like this:
```ts
type ChartSeriesOptions = Partial<Highcharts.SeriesOptionsType> & Record<string, any>
```

When we build the `Highcharts.Options` object, we have to explicitly cast the series Option to this type:
```ts
const options: Highcharts.Options = {
  chart: { type: 'bar' },
  series: [{
    data: [...],
    myProperty: 'myValue'
  }] as Array<ChartSeriesOptions> as Array<Highcharts.SeriesOptionsType>
}
```
This is accepted by the chart constructor (the series type is not required because it's taken from the chart type).
And if later we want to use the `myProperty` attribute, we cast the series object or the `userOptions` attribute to `ChartSeriesOptions`.

This is more or less how the old typings from DefinitelyTyped were working (but no explicitly casting was required).

> Agreed. We will try the outlined approach ASAP to make the introduction of custom series easier. Reference: #13374

Thanks! And sorry for putting multiple issues in one comment :)
